syntax = "proto3";

package autokitteh.user_code.v1;

import "autokitteh/user_code/v1/user_code.proto";
import "autokitteh/values/v1/values.proto";

message ExportsRequest {
  string file_name = 1;
}

message ExportsResponse {
  repeated string exports = 1;
  string error = 2;
}

message StartRequest {
  string entry_point = 1; // "main.py:on_event"
  Event event = 2;
}

message ExecuteRequest {
  bytes data = 1;
}

message ExecuteResponse {
  autokitteh.values.v1.Value result = 1;
  string error = 2;
  repeated Frame traceback = 3;
}

message StartResponse {
  string error = 1;
  repeated Frame traceback = 2;
}

message ActivityReplyRequest {
  autokitteh.values.v1.Value result = 1;
  string error = 2;
}

message ActivityReplyResponse {
  string error = 1;
}

message RunnerHealthRequest {}

message RunnerHealthResponse {
  string error = 1;
}

service RunnerService {
  // Get exports
  rpc Exports(ExportsRequest) returns (ExportsResponse) {}
  // Called at start of session
  rpc Start(StartRequest) returns (StartResponse) {} // initial session event
  // Execute a function in the runtime (skipped if it's a reply)
  rpc Execute(ExecuteRequest) returns (ExecuteResponse) {}
  // Reply from activity
  rpc ActivityReply(ActivityReplyRequest) returns (ActivityReplyResponse) {}

  rpc Health(RunnerHealthRequest) returns (RunnerHealthResponse) {}
}
